### ISSUE TITLE
figure out if we can reduce the number of allocations for a scheduled task

### ISSUE THREAD

**Issue by weissi** [apple/swift-nio#1316]
`eventLoop.scheduleTask(in:)` allocates a lot, I think #1315 shows that it's 9 (11 on macOS because of the `autoreleasepool`) allocations per scheduled task. I'm fairly sure we can lower this :)

**Comment by weissi** [comment]
<img width="129" alt="Screenshot 2021-05-14 at 2 42 54 pm" src="https://user-images.githubusercontent.com/624238/118279451-12163600-b4c3-11eb-9d26-390f9a0df5c5.png">

Seems to be 7 now, still quite a lot...

**Comment by weissi** [comment]
for reference, that's zoomed into a gRPC echo rpc

<img width="1062" alt="Screenshot 2021-05-14 at 1 45 36 pm" src="https://user-images.githubusercontent.com/624238/118279625-3e31b700-b4c3-11eb-88b4-498588e397b1.png">

**Comment by FranzBusch** [comment]
I just landed #2009, #2010 and #2011 which added baseline performance and allocation tests and reduced the allocation count for `execute` by 1 and for `scheduleTask` by 3 in total. There are only 2 more allocations that we can look at to optimise away for `scheduleTask`:

1. The `failFn` of `ScheduledTask` which I tried to do in #2012.
2. The `cancellationTask` of `Scheduled` that we are creating inside the `scheduleTask` method. 

Both of them are not trivial to get rid off and we would require significant changes to the internals of `SelectableEventLoop` or change the semantics of the APIs we vendor.

I would love to close this issue with these informations. cc @weissi @glbrntt
